### PR TITLE
Disable ptrace protection on ubuntu when EFA is enabled

### PR DESCRIPTION
**Pull request by demartinofra**
This is needed to use SHM transfer for intra-node communication

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: recipes/_efa_config.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+#
+# Cookbook Name:: aws-parallelcluster
+# Recipe:: _efa_config
```
Comment (enrico-usai) [inline recipes/_efa_config.rb]: AFAIK we are getting rid of the underscore prefix for the recipes.
If so I'd remove it from here and from the _efa_install too.

File: attributes/conditions.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+default['conditions']['efa_supported'] = (node['platform'] == 'centos' && node['platform_version'].to_i >= 7) \
```
Comment (enrico-usai) [inline attributes/conditions.rb]: Nice idea!

File: recipes/_efa_config.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    replace_or_add "disable ptrace protection" do
+      path "/etc/sysctl.d/10-ptrace.conf"
+      pattern "kernel.yama.ptrace_scope"
+      line "kernel.yama.ptrace_scope = 0"
+      notifies :run, 'execute[reload ptrace sysctl settings]', :immediately
+    end
+
+    execute "reload ptrace sysctl settings" do
+      action :nothing
+      command 'sysctl -p  /etc/sysctl.d/10-ptrace.conf'
```
Comment (enrico-usai) [inline recipes/_efa_config.rb]: minor: extra space

File: recipes/_efa_install.rb [hunk 4]
```diff
@@ -15,39 +15,41 @@
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
 
-efa_tarball = "#{node['cfncluster']['sources_dir']}/aws-efa-installer.tar.gz"
+if node['conditions']['efa_supported']
```
Comment (enrico-usai) [inline recipes/_efa_install.rb]: What about removing the nested if and use a guard like:
```ruby
return unless node['conditions']['efa_supported']
```

File: recipes/intel_mpi.rb [hunk 5]
```diff
@@ -15,46 +15,48 @@
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
 
-intelmpi_modulefile = "#{node['cfncluster']['modulefile_dir']}/intelmpi/#{node['cfncluster']['intelmpi']['version']}"
-intelmpi_installer = "#{node['cfncluster']['sources_dir']}/l_mpi_#{node['cfncluster']['intelmpi']['version']}.tgz"
+if node['conditions']['intel_mpi_supported']
```
Comment (enrico-usai) [inline recipes/intel_mpi.rb]: See my previous comment:
```ruby
return unless node['conditions']['intel_mpi_supported']
```

File: recipes/_efa_config.rb (hunk 1 above)
Comment (demartinofra) [reply recipes/_efa_config.rb]: done in a separate commit

File: recipes/_efa_config.rb (hunk 3 above)
Comment (demartinofra) [reply recipes/_efa_config.rb]: better than a linter :D

File: recipes/_efa_install.rb (hunk 4 above)
Comment (demartinofra) [reply recipes/_efa_install.rb]: good idea. done!

File: recipes/intel_mpi.rb (hunk 5 above)
Comment (demartinofra) [reply recipes/intel_mpi.rb]: done

File: recipes/efa_config.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the
+# License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+if node['platform'] == 'ubuntu'
```
Comment (enrico-usai) [inline recipes/efa_config.rb]: What about replacing the two nested ifs with a guard?! :D 
```ruby
return unless (node['platform'] == 'ubuntu' && node['cfncluster']['enable_efa'] == 'compute' && node['cfncluster']['cfn_node_type'] == 'ComputeFleet')
```

**Pull request merged**